samza-core/src/main/java/org/apache/samza/config/TaskConfig.java

```diff
@@ -0,0 +1,298 @@
+  // class name to use when sending offset checkpoints
+  public static final String CHECKPOINT_MANAGER_FACTORY = "task.checkpoint.factory";
+
+  public TaskConfig(Config config) {
+    super(config);
+  }
+
+  public Set<SystemStream> getInputStreams() {
+    Optional<String> inputStreams = Optional.ofNullable(get(INPUT_STREAMS));
+    if (!inputStreams.isPresent() || inputStreams.get().isEmpty()) {
+      return Collections.emptySet();
+    } else {
+      return Stream.of(inputStreams.get().split(","))
+          .map(systemStreamNames -> StreamUtil.getSystemStreamFromNames(systemStreamNames.trim()))
+          .collect(Collectors.toSet());
+    }
+  }
+
+  public long getWindowMs() {
+    return getLong(WINDOW_MS, DEFAULT_WINDOW_MS);
+  }
+
+  public long getCommitMs() {
+    return getLong(COMMIT_MS, DEFAULT_COMMIT_MS);
+  }
+
+  public Optional<String> getTaskClass() {
+    return Optional.ofNullable(get(TASK_CLASS));
+  }
+
+  public String getCommandClass(String defaultCommandClass) {
+    return get(COMMAND_BUILDER, defaultCommandClass);
+  }
+
+  public Optional<String> getMessageChooserClass() {
+    return Optional.ofNullable(get(MESSAGE_CHOOSER_CLASS_NAME));
+  }
+
+  public boolean getDropDeserializationErrors() {
+    return getBoolean(DROP_DESERIALIZATION_ERRORS, false);
+  }
+
+  public boolean getDropSerializationErrors() {
+    return getBoolean(DROP_SERIALIZATION_ERRORS, false);
+  }
+
+  public boolean getDropProducerErrors() {
+    return getBoolean(DROP_PRODUCER_ERRORS, false);
+  }
+
+  public Optional<Integer> getPollIntervalMs() {
+    return Optional.ofNullable(get(POLL_INTERVAL_MS)).map(Integer::parseInt);
+  }
+
+  public Optional<String> getIgnoredExceptions() {
+    return Optional.ofNullable(get(IGNORED_EXCEPTIONS));
+  }
+
+  public String getTaskNameGrouperFactory() {
+    Optional<String> taskNameGrouperFactory = Optional.ofNullable(get(GROUPER_FACTORY));
+    if (taskNameGrouperFactory.isPresent()) {
+      return taskNameGrouperFactory.get();
+    } else {
+      LOGGER.info(String.format("No %s configuration, using %s", GROUPER_FACTORY,
+          GroupByContainerCountFactory.class.getName()));
+      return GroupByContainerCountFactory.class.getName();
+    }
+  }
+
+  public int getMaxConcurrency() {
+    return getInt(MAX_CONCURRENCY, DEFAULT_MAX_CONCURRENCY);
+  }
+
+  public long getCallbackTimeoutMs() {
+    return getLong(CALLBACK_TIMEOUT_MS, DEFAULT_CALLBACK_TIMEOUT_MS);
+  }
+
+  public boolean getAsyncCommit() {
+    return getBoolean(ASYNC_COMMIT, false);
+  }
+
+  public long getMaxIdleMs() {
+    return getLong(MAX_IDLE_MS, DEFAULT_MAX_IDLE_MS);
+  }
+
+  /**
+   * Create the checkpoint manager
+   *
+   * @param metricsRegistry Registry of metrics to use. Can be null if not using metrics.
+   * @return CheckpointManager object if checkpoint manager factory is configured, otherwise empty.
+   */
+  public Optional<CheckpointManager> getCheckpointManager(MetricsRegistry metricsRegistry) {
```

Comment:
There are currently a few different places where object instantiations are done in config objects. Maybe we could have a separate clean-up ticket to handle all of those at once.
In the cases where we have similar flows (e.g. 1) get checkpoint manager factory class, 2) create the checkpoint manager factory instance, 3) create checkpoint manager), would you prefer to have a utility helper method instead of using the config object directly?